mixnet/node.py

```diff
@@ -1,107 +1,186 @@
 from __future__ import annotations
 
 import asyncio
-from contextlib import suppress
-from typing import Self, Tuple, TypeAlias
+import hashlib
+from typing import Awaitable, Callable, TypeAlias
 
-from cryptography.hazmat.primitives.asymmetric.x25519 import (
-    X25519PrivateKey,
-)
+from pysphinx.payload import DEFAULT_PAYLOAD_SIZE
 from pysphinx.sphinx import (
     Payload,
     ProcessedFinalHopPacket,
     ProcessedForwardHopPacket,
     SphinxPacket,
-    UnknownHeaderTypeError,
-)
-
-from mixnet.config import MixNodeConfig, NodeAddress
-from mixnet.poisson import poisson_interval_sec
-
-PacketQueue: TypeAlias = "asyncio.Queue[Tuple[NodeAddress, SphinxPacket]]"
-PacketPayloadQueue: TypeAlias = (
-    "asyncio.Queue[Tuple[NodeAddress, SphinxPacket | Payload]]"
 )
 
+from mixnet.config import GlobalConfig, NodeConfig
+from mixnet.packet import Fragment, MessageFlag, MessageReconstructor, PacketBuilder
 
-class MixNode:
-    """
-    A class handling incoming packets with delays
+NetworkPacket: TypeAlias = "SphinxPacket | bytes"
+NetworkPacketQueue: TypeAlias = "asyncio.Queue[NetworkPacket]"
+Connection: TypeAlias = NetworkPacketQueue
+BroadcastChannel: TypeAlias = "asyncio.Queue[bytes]"
 
-    This class is defined separated with the MixNode class,
-    in order to define the MixNode as a simple dataclass for clarity.
-    """
 
-    config: MixNodeConfig
-    inbound_socket: PacketQueue
-    outbound_socket: PacketPayloadQueue
-    task: asyncio.Task  # A reference just to prevent task from being garbage collected
+class Node:
+    config: NodeConfig
+    global_config: GlobalConfig
+    mixgossip_channel: MixGossipChannel
+    reconstructor: MessageReconstructor
+    broadcast_channel: BroadcastChannel
 
-    @classmethod
-    async def new(
-        cls,
-        config: MixNodeConfig,
-    ) -> Self:
-        self = cls()
+    def __init__(self, config: NodeConfig, global_config: GlobalConfig):
         self.config = config
-        self.inbound_socket = asyncio.Queue()
-        self.outbound_socket = asyncio.Queue()
-        self.task = asyncio.create_task(self.__run())
-        return self
+        self.global_config = global_config
+        self.mixgossip_channel = MixGossipChannel(
+            config.peering_degree, self.__process_sphinx_packet
+        )
+        self.reconstructor = MessageReconstructor()
+        self.broadcast_channel = asyncio.Queue()
+
+    async def __process_sphinx_packet(
+        self, packet: SphinxPacket
+    ) -> NetworkPacket | None:
+        try:
+            processed = packet.process(self.config.private_key)
+            match processed:
+                case ProcessedForwardHopPacket():
+                    return processed.next_packet
+                case ProcessedFinalHopPacket():
+                    await self.__process_sphinx_payload(processed.payload)
+        except Exception:
```

Comment:
Better to be clear with exceptions. What can fail here?

Reply:
235c703 I changed it to `ValueError` because `pysphinx` package doesn't raise well-defined exceptions yet. This is the case which the sphinx packet cannot be unwrapped by the private key of this node. In this case, the node should gossip it just as it is. I also updated the comment more clearly.